mrrc/pkgs/indexing.py

```diff
@@ -0,0 +1,257 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from botocore.exceptions import ClientError
+from mrrc.storage import S3Client
+from jinja2 import Template
+from treelib import Tree, Node
+import uuid
+import os
+from typing import List, Set
+
+
+class IndexHTML(object):
+    # object for holding index html file data
+    def __init__(self, title: str, header: str, items: Set[str]):
+        self.title = title
+        self.header = header
+        self.items = items
+
+    def generate_index_file_content(self) -> str:
+        template = Template(get_index_template())
+        return template.render(index=self)
+
+
+def get_index_template() -> str:
+    DEFAULT_INDEX_TEMPLATE = os.path.join(
+        os.environ["HOME"], ".mrrc/template/index.html.j2"
+    )
+    with open(DEFAULT_INDEX_TEMPLATE, encoding="utf-8") as file_:
+        return file_.read()
+
+
+def path_to_index(top_level: str, valid_paths: List[str], bucket: str):
+    repos = []
+    for path in valid_paths:
+        if path.startswith(top_level):
+            path = path.replace(top_level, '')
+            if path[0] == '/':
+                path = path[1:]
+            repos.append(path)
+
+    tree = tree_convert(repos)
+    index_files = html_convert(tree, '/', top_level, bucket)
+    return index_files
+
+
+def tree_convert(output):
+    tree = Tree()
+    temp_node = Node(identifier='root', tag="/", data="/")
+    tree.add_node(temp_node)
+
+    # example of line: net/java/jvnet-parent/4.0.0.redhat-3/jvnet-parent-4.0.0.redhat-3.pom
+    for line in output:
+        # escaped '/' could break logic here, improve in the future
+        paths = line.split('/')
+
+        paths = [p + '/' for p in paths[:-1]] + [paths[-1]]
+
+        id_holder = 'root'
+        for path in paths:
+            for child in tree.is_branch(id_holder):
+                if tree[child].tag == path:
+                    id_holder = child
+                    break
+            else:
+                temp_node = Node(identifier=uuid.uuid4(), tag=path, data=path)
+                tree.add_node(temp_node, parent=id_holder)
+                id_holder = temp_node.identifier
+                continue
+
+    return tree
+
+
+def get_update_list(repos: List[str], indexes: List[str], top_level: str, bucket: str) -> List[str]:
+
+    s3client = S3Client()
+
+    base_dir = top_level
+    if not top_level.endswith('/'):
+        base_dir += '/'
+    _repos = set(_.replace(base_dir, '') for _ in repos)
+    # filter out if only prod_key removed
+    _bucket = s3client._S3Client__get_bucket(bucket)
+    for repo in set(_repos):
+        file = _bucket.Object(repo)
+        exist = s3client._S3Client__file_exists(file)
```

Comment:
I'm really concerning about the performance for this code. Imaging there are millions of files, and will this brings millions request to s3 to check if these files exists? I think this is horrible.